interfaces/builtin/udisks2.go

```diff
@@ -431,24 +427,35 @@ func (iface *udisks2Interface) AppArmorConnectedPlug(spec *apparmor.Specificatio
 }
 
 func (iface *udisks2Interface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
-	if !release.OnClassic {
+	if !implicitSystemPermanentSlot(slot) {
 		spec.AddSnippet(udisks2PermanentSlotAppArmor)
 	}
 	return nil
 }
 
 func (iface *udisks2Interface) UDevPermanentSlot(spec *udev.Specification, slot *snap.SlotInfo) error {
-	if !release.OnClassic {
-		spec.AddSnippet(udisks2PermanentSlotUDev)
-		spec.TagDevice(`SUBSYSTEM=="block"`)
-		// # This tags all USB devices, so we'll use AppArmor to mediate specific access (eg, /dev/sd* and /dev/mmcblk*)
-		spec.TagDevice(`SUBSYSTEM=="usb"`)
+	if !implicitSystemPermanentSlot(slot) {
+		var udevFile string
+		slot.Attr("udev-file", &udevFile)
+		if udevFile != "" {
+			mountDir := slot.Snap.MountDir()
+			data, err := ioutil.ReadFile(filepath.Join(mountDir, udevFile))
```

Comment:
should we check that file is actually under mountDir and not outside the snap?

Reply:
Do you mean like `../../../../../../etc/passwd`?

Reply:
yes, not sure it is exploitable, but it still seems consistent with how tend to treat this kind of situations and might prevent silly mistakes (during development)

Reply:
An if we have symlinks in the path, we have resolve them with `mountDir` as a sysroot.

Reply:
I have not seen any function to resolve path in sysroot. Maybe I missed it. So I added one in osutil.

Reply:
Though, this is not a sysroot. It is expected to be mounted somewhere it does not know. Maybe I can change change the function to not allow resolve absolute symlinks or relative symlinks that would go out.

Reply:
I have changed a bit. I made the function verify that the path is valid rather than just make it safe.